cpp/src/experimental/scramble.cuh

```diff
@@ -0,0 +1,88 @@
+/* Copyright (C) 2009-2010 The Trustees of Indiana University.             */
+/*                                                                         */
+/* Use, modification and distribution is subject to the Boost Software     */
+/* License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at */
+/* http://www.boost.org/LICENSE_1_0.txt)                                   */
+/*                                                                         */
+/*  Authors: Jeremiah Willcock                                             */
+/*           Andrew Lumsdaine                                              */
+
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cassert>
+#include <cstdint>
+
+/* Apply a permutation to scramble vertex numbers; a randomly generated
+ * permutation is not used because applying it at scale is too expensive. */
+template <typename vertex_t>
+__device__ std::enable_if_t<sizeof(vertex_t) == 8, vertex_t> scramble(vertex_t value, size_t lgN)
+{
+  assert(std::is_unsigned<vertex_t>::value || lgN < 64);
+  assert(value >= 0);
+
+  constexpr uint64_t scramble_value0{606610977102444280};    // randomly generated
+  constexpr uint64_t scramble_value1{11680327234415193037};  // randomly generated
+
+  auto v = static_cast<uint64_t>(value);
```

Comment:
The common part of these 3 specialization could be factored out in a template function; e.g.,
```
template<typename precision_t, typename brev_wrapper_t>
__device__ decltype(auto) generate_v(precision_t const& value, 
 precision_t const& scramble0,
 precision_t const& scramble1, 
 precision_t const& mask0,
 precision_t const& mask1,
 size_t lgN,
 brev_wrapper_t brev)
{
  auto v = static_cast<precision_t>(value);
  v += scramble 0 + scramble 1;
  v *= (scramble0 | mask0);
  v = brev(v) >> (8*sizeof(precision_t) - lgN);
  v *= (scramble1| mask1);
  v = brev(v) >> (8*sizeof(precision_t) - lgN);
 return v;
}
``` 
Then in each specialization initialize the scrambles, masks, and the appropriate `brev` lambda; and pass all to `generate_v(...)`. Something like that. That way the underlying logic is factored out and only moving parts are kept separate.

Reply:
And `brev` can be constructed in-place as a `__device__` generic lambda (if possible... I think there might be some issues with `nvcc` compiling generic device lambdas). But, if possible, something like:
```
auto brev = [] __device__ (auto v){
 return __brevll(v);
};
//...
auto v = generate_v(..., brev);
```
Obviously, different specializations of `scramble(...)` would instantiate different versions of `brev`, accordingly, etc.

Reply:
OK, thanks for pointing this out, and I refactored the code to remove redundancy.